import_exports

juansqw opened the pull request:
Desarrolla funciones de importaciones y exportaciones

juansqw committed:
import_exports

Desarrolla funciones de importaciones y exportaciones

juansqw committed:
actualizacion

juansqw committed:
Update import_exports.R

juansqw committed:
Documentacion de detalles

juansqw committed:
Update import_exports.R

juansqw committed:
Update data.R

juansqw committed:
Merge branch 'main' into import-export

Johan-rosa commented on R/import_exports.R:
```diff
@@ -0,0 +1,194 @@
+    frecuencia,
+    choices = c("mensual", "trimestral", "anual"))
+
+  periodo <- c(2010:lubridate::year(Sys.Date()))
```
### Dos cosas:
1 sobre el cóndigo: `c()` no es necesario cuando usas el operador `:`
2 sobre el nombre: le pondría `years`

### alternativas

`years <- 2010:lubridate::year(Sys.Date())`
`years <- seq(2010, lubridate::year(Sys.Date()))`

Johan-rosa commented on R/import_exports.R:
```diff
@@ -0,0 +1,194 @@
+  url_descarga <- paste0(
+    "https://cdn.bancentral.gov.do/documents/estadisticas/",
+    "sector-externo/documents/Exportaciones_Mensuales_",
+    periodo, "_6.xls?v=1571253704905"
```
A los url le puedes quitar el argumento de la versión, `?v=1571253704905` en este caso

Johan-rosa commented on R/import_exports.R:
```diff
@@ -0,0 +1,194 @@
+  files_path <- purrr::map_chr(
+    periodo,
+    ~tempfile(pattern = "", fileext = ".xls")
+  )
```
No es necesario usar una purrr para crear varios archivos temporales, esa función es vectorizada.

`files_path <-tempfile(pattern = as.character(years), fileext = ".xls")`

juansqw committed:
Merge branch 'import-export' of https://github.com/Johan-rosa/databcrd into import-export

Johan-rosa commented on R/import_exports.R:
```diff
@@ -0,0 +1,194 @@
+    ~tempfile(pattern = "", fileext = ".xls")
+  )
+
+  dw_check <- purrr::possibly(utils::download.file, otherwise = -1) # nolint
```
Excelente uso de possibly para evitar que el error al intentar descargar un archivo que no existeEntiendo detenga la ejecusión de la función. 

También comprendo la lógica que quiere implementar usando `otherwhise = -1`, ahora bien, creo que hay mejores alternativas, más claras para mantener la función en el futuro. Paso a expandir mi opinión:

1. usar `otherwise = NA`
2. descargar los archivos usando `purrr::walk2`
3. usar `file.exists` para comprobar si efectivamente los archivos se descargaron y filtrar con eso

Te quedaría algo como sigue:

```r
  safe_download <- purrr::possibly(utils::download.file, otherwise = NA)

  purrr::walk2(
    url_descarga,
    files_path,
    \(url, file) safe_download(url, file, mode = "wb", quiet = TRUE)
  ) |>
    suppressWarnings()

  files_path <- files_path[file.exists(files_path)]
```

Johan-rosa commented on R/import_exports.R:
```diff
@@ -0,0 +1,194 @@
+    suppressWarnings(
+      exportaciones <- purrr::map(
+        files_path,
+        readxl::read_excel, sheet = 1,
```
`sheet = 1` no es necesario, ese es el argumento por defecto

Johan-rosa commented on R/import_exports.R:
```diff
@@ -0,0 +1,194 @@
+        readxl::read_excel, sheet = 1,
+        col_names = TRUE, skip = 8, na = "n.d.",
+        n_max = 70)  |>
+        stats::setNames(periodo[as.logical(unlist(descarga) + 1)])
```
En caso de que decidas usar el approach con `file.exist`

`stats::setNames(years[seq_along(files_path)])`

juansqw committed:
Update import_exports.R

juansqw committed:
Update global.R

Johan-rosa commented on R/import_exports.R:
```diff
@@ -0,0 +1,186 @@
+        tidyr::drop_na(x2) |>
+        dplyr::select(-x1, -x2, -dplyr::last_col()) |>
+        dplyr::bind_cols(exports_details) |>
+        tidyr::gather("mes", "valor_expor", -c(original_names, labels,
```
Usemos `tidyr::pivot_longer`

```r
tidyr::pivot_longer(
  names_to = "mes",
  values_to = "valor_expor",
  cols = -c(
    original_names, labels, short_names, categoria, nivel, direct_parent)
)
```

Johan-rosa commented on R/import_exports.R:
```diff
@@ -0,0 +1,186 @@
+        dplyr::mutate(mes = crear_mes(stringr::str_to_title(mes),
+                                      type = "text_to_number"),
+                      fecha = as.Date(paste(year, mes, "1", sep = "-")),
+                      trimestre = lubridate::quarter(fecha, with_year = TRUE))
```
1. La función `crear_mes` ha cambiado un poco, ya no necesita `str_to_title`.
2. usar `lubridate::make_date()` para crear `fecha`

```r
dplyr::mutate(
  mes = crear_mes(mes, type = "text_to_number"),
  fecha = lubridate::make_date(year, mes, "1"),
  trimestre = lubridate::quarter(fecha, with_year = TRUE)
)
```

Johan-rosa commented on tests/testthat/test-exports.R:
Lo recomendable es que los archivos con los test tengan el mismo nombre que el archivo en el que está la función, pero con un prefijo. En este caso deberías crear un solo archivo `test-imports-exports.R`

Lee sobre la función `usethis::use_test()`

Johan-rosa reviewed (changes requested):
Algunas sugerencias adicionales

juansqw committed:
Update functions

juansqw committed:
Update import_exports.R

juansqw committed:
agrega funciones extras

Agrega importaciones de petroleo y exportaciones zonas francas

juansqw committed:
Update import_exports.R

juansqw committed:
Update import_exports.R

Johan-rosa committed:
Update with changes in main

Merge branch 'main' into import-export

# Conflicts:
#	NAMESPACE
#	R/data.R
#	R/global.R

Johan-rosa committed:
Update imports-exports
